v3.1

Anarmard opened the pull request:
- Add ItemRequest (Controller, Service, Repository)
- Pageable for all GET methods
- Without tests

Anarmard committed:
v3.1
- Add ItemRequest (Controller, Service, Repository)
- Pageable for all GET methods
- Without tests

Anarmard committed:
v3.2
- Add ItemRequest (Controller, Service, Repository)
- Pageable for all GET methods
- Without tests

Anarmard committed:
v3.3
- Pass all Postman tests

Anarmard committed:
v3.4
- Add Tests: UserControllerTest and ItemControllerTest

Anarmard committed:
v3.5
- Add Tests: BookingControllerTest and ItemRequestControllerTest

Anarmard committed:
v3.51
- Add Tests: BookingControllerTest and ItemRequestControllerTest

alexache commented on src/test/java/ru/practicum/shareit/booking/BookingControllerTest.java:
```diff
@@ -0,0 +1,184 @@
+@WebMvcTest(controllers = BookingController.class)
+public class BookingControllerTest {
+    @Autowired
+    ObjectMapper mapper;
```
Модификаторов доступа не хватает

Anarmard replied:
Спасибо, поправил!

alexache commented on src/test/java/ru/practicum/shareit/booking/BookingControllerTest.java:
```diff
@@ -0,0 +1,184 @@
+                        Matchers.is(bookingResponseDto.getEnd().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME))))
+                .andExpect(jsonPath("$.status",
+                        Matchers.is(bookingResponseDto.getStatus().toString()), String.class))
+                .andExpect(status().isOk());
```
Суть подхода к тесту верная, только не помешает еще верификация в конце через Mockito.verify() на то что сервис был вызван и нужное количество раз

Anarmard replied:
Добавил, спасибо!

alexache commented on src/test/java/ru/practicum/shareit/booking/BookingControllerTest.java:
```diff
@@ -0,0 +1,184 @@
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .accept(MediaType.APPLICATION_JSON))
+                .andExpect(jsonPath("$.id", Matchers.is(bookingResponseDto.getId()), Long.class))
+                .andExpect(jsonPath("$.start",
```
(опционально) не обязательно прямо все-все поля проверять, можно ограничиться основными

Anarmard replied:
Хорошо)

alexache commented on src/test/java/ru/practicum/shareit/item/ItemControllerTest.java:
```diff
@@ -0,0 +1,204 @@
+    @Test
+    void deleteItemTest() throws Exception {
+        mvc.perform(delete("/items/1")
+                        .header("X-Later-User-Id", 1L)
```
Странный хедер

Anarmard replied:
)) исправил здесь и в самом контроллере (я так понимаю этот эндпоинт не нужен был, лишний).

Anarmard committed:
v3.6
- Add Tests: ItemServiceTest

Anarmard committed:
v3.61
- Add Tests: ItemServiceTest - continue

Anarmard committed:
v3.61
- Add Tests: ItemServiceTest - continue

Anarmard committed:
v3.62
- Add Tests: ItemServiceTest - continue

Anarmard committed:
Merge remote-tracking branch 'origin/add-item-requests' into add-item-requests

# Conflicts:
#	src/test/java/ru/practicum/shareit/item/ItemServiceImplTest.java

Anarmard committed:
v3.62
- Add Tests: ItemServiceTest - continue

Anarmard committed:
v3.7
- Add Additional Tests

Anarmard committed:
v3.7.1
- codestyle

Anarmard committed:
v3.7.2
- codestyle

Anarmard committed:
v3.8
- Add MapperImplTest

Anarmard committed:
v3.8.1
- Add MapperImplTest

Anarmard committed:
v3.8.2
- Add MapperImplTest

Anarmard committed:
v3.8.3
- Add MapperImplTest

Anarmard committed:
v3.9
- findAllByBookerAndStartBeforeAndEndAfterOrderByStartAsc
раньше было на конце Desc

Anarmard committed:
v3.10
- ошибка в БД itemRequest / название столбца

felixoff commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -39,18 +39,22 @@ public BookingResponseDto getBooking(@RequestHeader("X-Sharer-User-Id") Long use
-                                                          defaultValue = "ALL") String state) {
-        return bookingService.getAllBooking(userId, state);
+    public List<BookingResponseDto> getAllBooking(
+            @RequestHeader("X-Sharer-User-Id") Long userId,
```
Хедеры можно вынести в отдельную переменную для избежания дубляжей

Anarmard replied:
ОК. Надеюсь я верно сделал то, что ты просил (в переменную засунул только название заголовка... как весь заголовок через переменную сделать, не знаю и не смог найти такую инфу): 
private static final String USERID = "X-Sharer-User-Id";

felixoff commented on src/main/java/ru/practicum/shareit/booking/dto/BookingCreateRequestDto.java:
```diff
@@ -2,23 +2,28 @@
 
     @FutureOrPresent
+    @NotNull
     private LocalDateTime start; // дата и время начала бронирования
```
Вместо множественных private лучше использовать @FieldDefaults

Anarmard replied:
Спасибо большое, не знал про такую аннотацию! Во всех dto и model указал ее.

felixoff commented on src/test/java/ru/practicum/shareit/request/ItemRequestServiceImplTest.java:
```diff
@@ -0,0 +1,173 @@
+                .thenReturn(Optional.empty());
+        Exception e0 = Assertions.assertThrows(NotFoundException.class,
+                () -> itemRequestService.addNewItemRequest(itemRequestDto, 4L));
+        Assertions.assertEquals("User with such ID does not exist", e0.getMessage());
```
Лучше указывать в комментариях пользователю конкретный идентификатор

Anarmard replied:
Понял, спасибо. Поправил, где применимо.

Anarmard committed:
v3.11
- USERID = "X-Sharer-User-Id"
- не such ID, а указать сам ID
- @fielddefaults
